Upgrade operator-sdk v0.16.0

rcernich opened the pull request:
The bulk of the changes deal with upgrading k8s.io/controller-runtime from v0.1.x to v0.4.x.  The changes also collapse the validating/mutating webhook controllers into a single controller that handles caBundle updates for all webhooks.

This is divided into two commits to separate the vendor changes from the actual code changes.

rcernich committed:
update vendor for operator-sdk v0.15.2

Signed-off-by: rcernich <rcernich@redhat.com>

rcernich committed:
migrate code to work with operator-sdk v0.15.2

Signed-off-by: rcernich <rcernich@redhat.com>

rcernich commented:
This PR is good to merge, should we decide to upgrade operator-sdk for Maistra 1.1.  Note, the only issue at present is the reliance on service-ca to create the serving secret for the webhook server.  This causes the initial startup of the operator to fail (i.e. bounce).  Subsequent startups succeed and the webhook server starts normally.  (In other words, the operator will restart once when it is first installed.)

luksa commented:
> Note, the only issue at present is the reliance on service-ca to create the serving secret for the webhook server. This causes the initial startup of the operator to fail (i.e. bounce). Subsequent startups succeed and the webhook server starts normally. (In other words, the operator will restart once when it is first installed.)

I think we should introduce another controller that ensures the service and the webhookconfig objects are created (and recreated whenever they are deleted). The same controller could also watch for the serving-cert secret and write the files locally. This way, there is no need for the operator to restart. (note that running the operator locally is now harder because you need to extract the secret to /tmp manually).

luksa commented:
Forgot to say that the PR looks good otherwise. I've tested the operator manually and it appears to work.

I'm OK with this going in. I can create the new controller in a follow-up PR. 

Also, upgrading controller-runtime is a good idea, as it will allow us to do some things that weren't possible before: like forcing the update of the smcp/smmr/smm status (preventing conflicts). This will make the code simpler and ensure that the status is _always_ updated - even if users modify the object while the controller is reconciling it.

rcernich commented:
Yeah, the issue with the webhook is compounded by the fact that we can't read objects before the cache is started.  I suppose we could create our own client, using the config on the manager object.  It would be great if we could wait until the secret is available, to prevent a restart.  (The issue is that the webhook server fails to start if the key is missing.  The key gets mounted after the secret is created, which is done via service-ca reacting to the creation of the service.)

Also, the newer controller-runtime includes a `Patch()` method on the `Client`, which would help simplify the patching code (maybe).  I didn't tackle anything beyond what was required to upgrade though.

rcernich commented:
Forgot to mention that this work includes a refactoring of the webhook cabundle controllers, merging them into a single controller that watches both validating and mutating webhooks.  The new version of that controller is used to keep the cert on the operator's webhook up to date.

rcernich commented:
I've rebased and tested again, and everything looks good.  I've added some code to wait for the secret to become available before starting the controller manager, which allows the webhook server to start without error (because the secret with the tls.key has been mounted).

@knrc, after talking with @luksa today, we think we should merge this for 1.1.  For the record, these changes do add a dependency on the service-ca operator, which is a core cluster operator and should always be present. (By dependency, I mean we rely on it to generate a secret that is used to configure tls on the webhook server.  This used to be done by the controller-runtime framework, but that functionality has long since been removed.)

knrc commented:
@rcernich It's too big and risky at this stage, we're supposed to be finishing things off for the release and not introducing risky changes.

rcernich commented:
On hold until maistra-1.2 branch is created.

rcernich committed:
migrate code to work with operator-sdk v0.16.0

Signed-off-by: rcernich <rcernich@redhat.com>

rcernich committed:
update vendor for operator-sdk v0.16.0

Signed-off-by: rcernich <rcernich@redhat.com>

rcernich commented:
This should be good to go now.  Note, after rebasing, I updated to v0.16.0.  I tried v0.17.0, but there were issues with the dependencies, so I bailed.  I think it was the result of using 1.16 k8s with 1.17 client-go.  The migration docs didn't mention anything about updating the k8s dependencies, so I bailed out.

luksa commented on deploy/maistra-operator.yaml:
```diff
@@ -512,3 +533,7 @@ spec:
+      - name: webhook-tls-volume
+        secret:
+          secretName: maistra-operator-serving-cert
+          optional: true # this won't be available untile service-ca creates the secret
```
typo in "untile"

rcernich replied:
fixed

luksa commented on deploy/servicemesh-operator.yaml:
```diff
@@ -517,3 +538,7 @@ spec:
+      - name: webhook-tls-volume
+        secret:
+          secretName: maistra-operator-serving-cert
+          optional: true # this won't be available untile service-ca creates the secret
```
typo in "untile"

rcernich replied:
fixed

luksa commented on cmd/manager/main.go:
```diff
@@ -142,7 +155,32 @@ func main() {
+	// Get the namespace the operator is currently deployed in.
+	operatorNs, err := k8sutil.GetOperatorNamespace()
+	if err != nil {
+		if errors.Is(err, k8sutil.ErrRunLocal) {
```
This also needs to check for `ErrNoNamespace`.

rcernich replied:
This was copied directly from the operator-sdk upgrade docs.  My understanding is that the service isn't created if running locally, with `OSDK_FORCE_RUN_MODE=local`.  If `ErrNoNamespace` is returned, it means the operator is not running locally and is not running in a namespace (?).

I've updated `common.GetOperatorNamesapce()` to handle the `ErrRunLocal` error instead of `ErrNoNamespace` error, so local debugging should work if you set the env `OSDK_FORCE_RUN_MODE=local`

rcernich replied:
Looks like `ErrNoNamespace` was required for the unit tests, so I've added that back (or'd with `ErrRunLocal`).

luksa reviewed (approved):
LGTM. Let's get this in as soon as possible, as it will open up a few things I've been waiting to implement.

rcernich commented:
@dgn, could you give a review?

dgn commented:
This segfaults on my cluster when reconciling the SMCP, haven't figured out what's going on exactly yet though

```
E0511 16:16:36.415797       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 1987 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x1a68360, 0x3047130)
	/home/dgrimm/dev/go/src/github.com/maistra/istio-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/home/dgrimm/dev/go/src/github.com/maistra/istio-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:48 +0x82
panic(0x1a68360, 0x3047130)
	/usr/lib/golang/src/runtime/panic.go:679 +0x1b2
github.com/maistra/istio-operator/pkg/controller/common/helm.(*PatchFactory).CreatePatch(0x0, 0x1f95780, 0xc000f73760, 0x1f95780, 0xc000f73750, 0xc000221451, 0x9, 0x1f95780, 0xc000f73760)
	/home/dgrimm/dev/go/src/github.com/maistra/istio-operator/pkg/controller/common/helm/patch.go:46 +0xb2
github.com/maistra/istio-operator/pkg/controller/common/helm.(*ManifestProcessor).processObject(0xc0012b74f0, 0x1fcb440, 0xc0029749f0, 0xc000f73750, 0x1ce4105, 0x9, 0x1fcb440, 0xc0029749f0)
	/home/dgrimm/dev/go/src/github.com/maistra/istio-operator/pkg/controller/common/helm/manifestprocessing.go:150 +0x6a8
github.com/maistra/istio-operator/pkg/controller/common/helm.(*ManifestProcessor).ProcessManifest(0xc0012b74f0, 0x1fcb440, 0xc002974480, 0xc002926390, 0x2b, 0xc00292e5a0, 0x10e, 0xc003b556b0, 0x1ce4105, 0x9, ...)
	/home/dgrimm/dev/go/src/github.com/maistra/istio-operator/pkg/controller/common/helm/manifestprocessing.go:81 +0x792
github.com/maistra/istio-operator/pkg/controller/common/helm.(*ManifestProcessor).ProcessManifests(0xc0012b74f0, 0x1fcb440, 0xc0020a6360, 0xc0025d7520, 0x5, 0xa, 0x1ce4105, 0x9, 0x0, 0x0)
	/home/dgrimm/dev/go/src/github.com/maistra/istio-operator/pkg/controller/common/helm/manifestprocessing.go:48 +0x311
github.com/maistra/istio-operator/pkg/bootstrap.internalInstallCNI(0x1fcb440, 0xc0020a6360, 0x1fe99e0, 0xc000a8fe00, 0xc000b96e01, 0x0, 0x0, 0x0, 0xc001d0fb30, 0x0)
	/home/dgrimm/dev/go/src/github.com/maistra/istio-operator/pkg/bootstrap/cni.go:51 +0x682
github.com/maistra/istio-operator/pkg/bootstrap.InstallCNI(...)
	/home/dgrimm/dev/go/src/github.com/maistra/istio-operator/pkg/bootstrap/cni.go:20
github.com/maistra/istio-operator/pkg/controller/servicemesh/controlplane.(*controlPlaneInstanceReconciler).Reconcile(0xc002182e70, 0x1fcb440, 0xc0020a6360, 0x0, 0x0, 0x0, 0x0)
	/home/dgrimm/dev/go/src/github.com/maistra/istio-operator/pkg/controller/servicemesh/controlplane/reconciler.go:217 +0x1b83
github.com/maistra/istio-operator/pkg/controller/servicemesh/controlplane.(*ControlPlaneReconciler).Reconcile(0xc0006488c0, 0xc0014c97ba, 0x2, 0xc0014c97a0, 0xd, 0xc00157bc00, 0x0, 0x0, 0x0)
	/home/dgrimm/dev/go/src/github.com/maistra/istio-operator/pkg/controller/servicemesh/controlplane/controller.go:227 +0x867
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000c1d740, 0x1ae08a0, 0xc000d50220, 0x0)
	/home/dgrimm/dev/go/src/github.com/maistra/istio-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:256 +0x162
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000c1d740, 0xc0034f1c00)
	/home/dgrimm/dev/go/src/github.com/maistra/istio-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:232 +0xcb
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc000c1d740)
	/home/dgrimm/dev/go/src/github.com/maistra/istio-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:211 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0032d51f0)
	/home/dgrimm/dev/go/src/github.com/maistra/istio-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0032d51f0, 0x3b9aca00, 0x0, 0x1, 0xc0007f9c80)
	/home/dgrimm/dev/go/src/github.com/maistra/istio-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc0032d51f0, 0x3b9aca00, 0xc0007f9c80)
	/home/dgrimm/dev/go/src/github.com/maistra/istio-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	/home/dgrimm/dev/go/src/github.com/maistra/istio-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:193 +0x328
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x1852cc2]
```

luksa commented:
Rob needs to rebase the PR to pull in #428 (or you can just cherrypick it on top when testing the PR).

rcernich commented:
I'll rebase and push anew.

rcernich commented:
pushed a rebase against latest maistra-1.2 branch
